kaitai/kaitaistream.cpp

```diff
@@ -425,41 +432,106 @@ std::string kaitai::kstream::bytes_terminate(std::string src, char term, bool in
 // ========================================================================
 
 std::string kaitai::kstream::process_xor_one(std::string data, uint8_t key) {
+    if (key == 0)
+        return data;
+
     size_t len = data.length();
     std::string result(len, ' ');
 
-    for (size_t i = 0; i < len; i++)
+    for (size_t i = 0; i < len; i++) {
         result[i] = data[i] ^ key;
+    }
 
     return result;
 }
 
 std::string kaitai::kstream::process_xor_many(std::string data, std::string key) {
     size_t len = data.length();
     size_t kl = key.length();
+    if (len == 1)
+        return process_xor_one(data, key[0]);
+
     std::string result(len, ' ');
 
-    size_t ki = 0;
     for (size_t i = 0; i < len; i++) {
-        result[i] = data[i] ^ key[ki];
-        ki++;
-        if (ki >= kl)
-            ki = 0;
+        result[i] = data[i] ^ key[i % kl];
```

Comment:
I came to understanding, although I have to admit that I dont remember where exactly I learned this, so dont ask me for reference or benchmark data, that modulo operation is as cheap as any other "atomic" operation. 
You see, in Python which is my primary lang, there is literally no difference in perfomance between >> or | or + or * or / or %. That is because a simple operator call involves a lot of red tape, checking if both operands/objects are integers or floats, etc. The actual "C operation" is one of many instructions. 
In lower level langs like C# which is my previous primary language, the difference is a bit more pronounced but not as much as people make you believe. At university they teach you that bitwise ops are cheaper than integer addition then multiplication etc. In reality as benchmarks show, the difference is so small, that you just dont care.
So assuming that this theory holds for C++ as well, single op (modulo) is better than several ops (addition, if, branch misprediction).

Reply:
Oh its also consistent with the other runtimes implementations. :)

Reply:
I agree that we need a benchmark to know for sure. Please standby.